README.md
# jquery-ui-durationslider

Bind text field and jquery-ui slider for duration input.

## Dependencies

* jquery-ui

## Installation

Install from npm:

    $ npm install @kanety/jquery-ui-durationslider --save

## Usage

Make a text field and `div` elements for sliders:

```html
<input id="text" type="text" value="10:08" style="width: 5em;" />
<div id="hour" style="width: 150px;"></div>
<div id="minute" style="width: 150px;"></div>
```

Then:

```javascript
$('#text').durationslider({
  h: { elem: '#hour' },
  m: { elem: '#minute' }
});
```

If you want a slider for seconds:

```javascript
$('#text').durationslider({
  h: { elem: '#hour' },
  m: { elem: '#minute' },
  s: { elem: '#second' }
});
```

Customize max value and step value:

```javascript
$('#text').durationslider({
  h: { elem: '#hour', max: 48 },
  m: { elem: '#minute', step: 5 }
});
```

Customize duration format:

```javascript
$('#text').durationslider({
  format: 'd [DAYS] hh.mm'
});
```

* d: days
* h: hours
* m: minutes
* s: seconds
* []: raw text

Enable mousewheel support (only modern browsers):

```javascript
$('#text').durationslider({
  mousewheel: true
});
```

## License

The library is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# CHANGELOG

## 0.4.1

* Fix error on empty value.

## 0.4.0

* Remove moment dependency.

## 0.3.0

* Export class for convenience.
* Refactoring.

## 0.2.0

* Simplify slider options.
* Support day's slider.
* Support mouse wheel event.
* Refactoring.

## 0.1.0

* First release.


test/jquery-ui-durationslider-spec.js
describe('jquery-durationslider', function() {
  beforeEach(function() {
    document.body.innerHTML = __html__['index.html'];
  });

  it('has basic sliders', function() {
    var $ex = $('#basic');
    var $hour = $('#basic_hour');
    var $minute = $('#basic_minute');
    $ex.durationslider({
      h: { elem: $hour },
      m: { elem: $minute }
    });

    expect($ex.val()).toEqual('10:08');
    expect($hour.slider('value')).toEqual(10);
    expect($minute.slider('value')).toEqual(8);

    $hour.slider('value', 11).trigger('slide', { value: 11 });
    $minute.slider('value', 9).trigger('slide', { value: 9 });
    expect($ex.val()).toEqual('11:09');
    expect($hour.slider('value')).toEqual(11);
    expect($minute.slider('value')).toEqual(9);
  });

  it('includes sliders for days and seconds', function() {
    var $ex = $('#extend');
    var $day = $('#extend_day');
    var $hour = $('#extend_hour');
    var $minute = $('#extend_minute');
    var $second = $('#extend_second');
    $ex.durationslider({
      format: 'd [DAYS] hh:mm:ss',
      d: { elem: $day },
      h: { elem: $hour },
      m: { elem: $minute },
      s: { elem: $second }
    });

    expect($ex.val()).toEqual('1 DAYS 10:08:59');
    expect($day.slider('value')).toEqual(1);
    expect($hour.slider('value')).toEqual(10);
    expect($minute.slider('value')).toEqual(8);
    expect($second.slider('value')).toEqual(59);
  });

  it('handles mousewheel', function() {
    // IE can't handle WheelEvent
    var ua = window.navigator.userAgent.toLowerCase();
    if (ua.indexOf('msie') != -1 || ua.indexOf('trident') != -1) {
      return;
    }

    var $ex = $('#wheel');
    var $hour = $('#wheel_hour');
    var $minute = $('#wheel_minute');
    $('#wheel').durationslider({
      h: { elem: $hour },
      m: { elem: $minute },
      mousewheel: true
    });

    $hour.find('.ui-slider-handle').addClass('ui-state-focus');
    $minute.find('.ui-slider-handle').addClass('ui-state-focus');

    var event = new WheelEvent('wheel', { deltaX: 0, deltaY: -1, deltaZ: 0 });
    window.dispatchEvent(event);

    expect($ex.val()).toEqual('01:01');
  });
});


karma.conf.js
// Karma configuration
// Generated on Fri Nov 23 2018 10:20:11 GMT+0900 (JST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],


    // list of files / patterns to load in the browser
    files: [
      'node_modules/jquery/dist/jquery.min.js',
      'node_modules/jquery-ui-dist/jquery-ui.min.js',
      'dist/*.js',
      'index.html',
      'test/**/*spec.js'
    ],


    // list of files / patterns to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
        'src/*.js' : ['coverage'],
        'index.html': ['html2js']
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress', 'coverage'],

    coverageReporter: {
        type: 'html',
        dir: 'coverage/'
    },

    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: false,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: (function() {
      if (process.platform === "win32") {
        return ['Chrome', 'Firefox', 'IE', 'Edge']
      } else {
        return ['ChromeHeadless', 'FirefoxHeadless']
      }
    })(),


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true,

    // Concurrency level
    // how many browser should be started simultaneous
    // concurrency: Infinity
    concurrency: 1
  })
}


webpack.config.js
const webpack = require("webpack");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");
const UglifyJsPlugin = require("uglifyjs-webpack-plugin");
const OptimizeCSSAssetsPlugin = require("optimize-css-assets-webpack-plugin");

module.exports = {
  entry: {
    "jquery-ui-durationslider": "./src/jquery-ui-durationslider.js"
  },

  output: {
    path: __dirname + "/dist",
    filename: "[name].js",
    publicPath: "/dist"
  },

  externals: {
    jquery: "jQuery"
  },

  plugins: [
    new MiniCssExtractPlugin({
      filename: "[name].css"
    })
  ],

  optimization: {
    minimizer: [
      new UglifyJsPlugin({
        cache: true,
        parallel: true
      }),
      new OptimizeCSSAssetsPlugin({})
    ]
  },

  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
          options: {
            presets: ["@babel/preset-env"]
          }
        }
      }, {
        test: /\.scss$/,
        use: [
          MiniCssExtractPlugin.loader,
          "css-loader",
          "sass-loader"
        ],
      }
    ]
  },

  watchOptions: {
    poll: 1000
  },

  devServer: {
    host: "0.0.0.0",
    port: 3000,
    disableHostCheck: true
  }
};


src/wheel-handler.js
import { NAMESPACE } from './consts';

export default class WheelHandler {
  static bind() {
    let events = $._data(window, 'events');
    if (!events || !events.wheel || !events.wheel.some(e => e.namespace == NAMESPACE)) {
      $(window).on(`wheel.${NAMESPACE}`, (e) => {
        WheelHandler.handle(e);
      });
    }
  }

  static unbind() {
    $(window).off(`wheel.${NAMESPACE}`);
  }

  static handle(e) {
    $(document).find('.ui-slider-handle.ui-state-focus').each((i, elem) => {
      let $slider = $(elem).parent();
      let instance = $slider.data(NAMESPACE);
      if (instance && instance.options.mousewheel) {
        e.preventDefault();
        WheelHandler.wheelMoved(instance, $slider, WheelHandler.isUp(e));
      }
    })
  }

  static isUp(e) {
    return e.originalEvent && e.originalEvent.deltaY && e.originalEvent.deltaY > 0;
  }

  static wheelMoved(instance, $slider, up) {
    let value = $slider.slider('value');
    let option = $slider.slider('option');

    if (up) {
      value = Math.max(option.min, value - option.step);
    } else {
      value = Math.min(option.max, value + option.step);
    }

    $slider.slider('value', value);
    instance.sliderChanged();
  }
}


src/durationslider/mousewheel-handler.js
'use strict';

import { NAMESPACE } from '../consts';

export default class MousewheelHandler {
  static bind() {
    let events = $._data(window, 'events');
    if (!events || !events.wheel || !events.wheel.some(e => e.namespace == NAMESPACE)) {
      $(window).on(`wheel.${NAMESPACE}`, MousewheelHandler.handle);
    }
  }

  static unbind() {
    $(window).off(`wheel.${NAMESPACE}`);
  }

  static handle(e) {
    $(document).find('.ui-slider-handle.ui-state-focus').each(function(i, elem) {
      let $slider = $(elem).parent();
      let instance = $slider.data(`${NAMESPACE}-input`).data(NAMESPACE);
      if (instance && instance.options.mousewheel) {
        e.preventDefault();
        instance.wheelMoved($slider, MousewheelHandler.isUp(e));
      }
    })
  }

  static isUp(e) {
    return e.originalEvent.deltaY && e.originalEvent.deltaY < 0;
  }
}


src/durationslider.js
import { NAMESPACE } from './consts';
import WheelHandler from './wheel-handler';

const DEFAULTS = {
  d: {
    elem: null,
    min: 0,
    max: 10,
    step: 1
  },
  h: {
    elem: null,
    min: 0,
    max: 23,
    step: 1
  },
  m: {
    elem: null,
    min: 0,
    max: 59,
    step: 1
  },
  s: {
    elem: null,
    min: 0,
    max: 59,
    step: 1
  },
  format: null,
  mousewheel: false
};

export default class Durationslider {
  constructor(input, options = {}) {
    this.options = $.extend(true, {}, DEFAULTS, options);

    this.$input = $(input);
    this.init();
  }

  init() {
    this.$sliders = [];

    ['d', 'h', 'm', 's'].forEach((type) => {
      let option = this.options[type];
      if (!option.elem) {
        return;
      }

      let $slider = $(option.elem).addClass(NAMESPACE).data(NAMESPACE, this);
      $slider.slider({ min: option.min, max: option.max, step: option.step, type: type });
      this.$sliders.push($slider);
    });

    if (!this.options.format) {
      this.options.format = this.defaultFormat();
    }

    this.textChanged();
    this.bind();
  }

  defaultFormat() {
    return this.$sliders.map(($slider) => {
      let type = $slider.slider('option', 'type');
      return `${type}${type}`;
    }).join(':');
  }

  bind() {
    this.$input.on(`input.${NAMESPACE}`, (e) => {
      this.textChanged();
    }).on(`blur.${NAMESPACE}`, (e) => {
      if (this.$input.val()) {
        this.sliderChanged();
      }
    });

    this.$sliders.forEach(($slider) => {
      $slider.on(`slide.${NAMESPACE}`, (e, ui) => {
        $(e.target).slider('value', ui.value);
        this.sliderChanged();
      });
    });

    if (this.options.mousewheel) {
      WheelHandler.bind();
    }
  }

  unbind() {
    this.$input.off(`.${NAMESPACE}`);

    this.$sliders.forEach(($slider) => {
      $slider.off(`.${NAMESPACE}`);
    });
  }

  dhms() {
    return Durationslider.toDHMS(this.$input.val(), this.options.format);
  }

  seconds() {
    return this.$sliders.map(($slider) => {
      return Durationslider.toSecond($slider.slider('option', 'type'), $slider.slider('value'));
    }).reduce((x, y) => {
      return x + y;
    });
  }

  textChanged() {
    let time = this.dhms();
    this.$sliders.forEach(($slider) => {
      $slider.slider('value', time[$slider.slider('option', 'type')]);
    });
  }

  sliderChanged() {
    let text = Durationslider.toText(this.seconds(), this.options.format);
    if (this.$input.val() != text) {
      this.$input.val(text).trigger('change');
    }
  }

  static toDHMS(text, format) {
    let values = text.match(/\d+/g) || [];
    let formats = format.replace(/\[.*?\]/g, '').match(/d+|h+|m+|s+/g) || [];

    let time = { d: 0, h: 0, m: 0, s: 0 };
    for (let i=0; i<Math.min(values.length, formats.length); i++) {
      let val = values[i];
      let fmt = formats[i];
      if (fmt.match(/[d]/)) {
        time.d += Number(val);
      } else if (fmt.match(/[h]/)) {
        time.h += Number(val);
      } else if (fmt.match(/[m]/)) {
        time.m += Number(val);
      } else if (fmt.match(/[s]/)) {
        time.s += Number(val);
      }
    }
    return time;
  }

  static toSecond(type, value) {
    switch (type) {
    case 'd':
      return value * 3600 * 24;
    case 'h':
      return value * 3600;
    case 'm':
      return value * 60;
    case 's':
      return value;
    }
    return 0;
  }

  static toText(seconds, format) {
    let formats = format.match(/\[.+?\]|d+|h+|m+|s+|./g) || [];

    let time = { d: 0, h: 0, m: 0, s: 0 };
    if (formats.some((fmt) => fmt.match(/^d+$/))) {
      time.d = Math.floor(seconds / (3600 * 24));
    }
    if (formats.some((fmt) => fmt.match(/^h+$/))) {
      time.h = Math.floor((seconds - (time.d * 3600 * 24)) / 3600);
    }
    if (formats.some((fmt) => fmt.match(/^m+$/))) {
      time.m = Math.floor((seconds - (time.d * 3600 * 24) - (time.h * 3600)) / 60);
    }
    if (formats.some((fmt) => fmt.match(/^s+$/))) {
      time.s = seconds - (time.d * 3600 * 24) - (time.h * 3600) - (time.m * 60);
    }

    let text = '';
    formats.forEach((fmt) => {
      let part;
      if (fmt.match(/^\[.+\]$/)) {
        part = fmt.slice(1, -1);
      } else if (fmt.match(/^d+$/)) {
        part = Durationslider.padding(time.d.toString(), fmt.length);
      } else if (fmt.match(/^h+$/)) {
        part = Durationslider.padding(time.h.toString(), fmt.length);
      } else if (fmt.match(/^m+$/)) {
        part = Durationslider.padding(time.m.toString(), fmt.length);
      } else if (fmt.match(/^s+$/)) {
        part = Durationslider.padding(time.s.toString(), fmt.length);
      } else {
        part = fmt;
      }
      text += part;
    });
    return text;
  }

  static padding(text, length) {
    for (let i=0; i<length - text.length; i++) {
      text = '0' + text;
    }
    return text;
  }

  static getDefaults() {
    return DEFAULTS;
  }

  static setDefaults(options) {
    $.extend(true, DEFAULTS, options);
  }
}
